### PR TITLE
feat: add rate limit implementation

### PR DESCRIPTION
**Pull request by mattsse**
Add a rate limit implementation.

This can be used to enforce rate limits like max `10 req / s`

Will be used to rate limit DNS lookups

### CODE REVIEW

File: crates/net/common/src/ratelimit.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+        limit.tick();
+
+        poll_fn(|cx| {
+            assert!(limit.poll_ready(cx).is_pending());
+            Poll::Ready(())
+        })
+        .await;
+
+        tokio::time::sleep(limit.rate.duration).await;
```
Comment (onbjerg) [inline crates/net/common/src/ratelimit.rs:143]:
```suggestion
        tokio::time::advance(limit.rate.duration).await;
```

File: crates/net/common/src/ratelimit.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    /// Create a new rate limiter
+    pub fn new(rate: Rate) -> Self {
+        let until = tokio::time::Instant::now();
+        let state = State::Ready { until, remaining: rate.limit() };
+
+        RateLimit { rate, state, sleep: Box::pin(tokio::time::sleep_until(until)) }
+    }
+
+    /// Checks if the [RateLimit] is ready to handle a new call
+    pub fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()> {
```
Comment (onbjerg) [inline crates/net/common/src/ratelimit.rs:31]: Is the intention that you call `poll_ready(cx)` every time before calling `tick`? I.e. let's say I wanted to send a request, I would need to do:

```
if rate_limiter.poll_ready(cx).is_ready() {
  // send req
  rate_limiter.tick();
}
```

And this is only usable in a hand-rolled future, right? Should we not have an API for things that do not have explicit access to `cx`?

Comment (mattsse) [reply crates/net/common/src/ratelimit.rs:31]: yes,

only for future::poll, but can use `poll_fn` though, unsure about async code yet

**Review by onbjerg — APPROVED** [review]

**Comment by mattsse** [comment]
added `async wait`

there's no `tokio::time::advance` though

**Pull request merged**